rockchip: add kernel 6.12 support

1715173329 opened the pull request:
Run-tested on NanoPi R3S.

1715173329 committed:
rockchip: refresh kernel patches

Remove upstreamed patches and refresh remaining ones.

Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>
Link: https://github.com/openwrt/openwrt/pull/18683
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

1715173329 committed:
rockchip: refresh kernel configs

Refresh kernel configs with `make kernel_oldconfig CONFIG_TARGET=subtarget`.

Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>
Link: https://github.com/openwrt/openwrt/pull/18683
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

1715173329 committed:
rockchip: add kernel 6.12 as testing kernel

Add kernel 6.12 as testing kernel.

Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>
Link: https://github.com/openwrt/openwrt/pull/18683
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

kmarzic commented:
Hi,
Question:
Bug with GMAC and VLANs is fixed with the new kernel (https://github.com/openwrt/openwrt/issues/16162)?

1715173329 commented:
iirc no, feel free to test.

Fil010 commented:
Tested-by: Francisco G Luna <frangonlun@gmail.com> #lunzn, Fastrhino-r68s

jmccrohan commented:
This PR runs on my Nano-Pi R6S, but performance is absolutely terrible. High packet loss which makes it essentially unusable. I had to roll back to previous 6.6 based image. See ping excerpt below. Nothing obvious in kernel log. Let me know if you need me to do any further testing.

```
root@OpenWrt:~# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: seq=0 ttl=118 time=9.212 ms
64 bytes from 8.8.8.8: seq=15 ttl=118 time=13.791 ms
64 bytes from 8.8.8.8: seq=21 ttl=118 time=9.047 ms
64 bytes from 8.8.8.8: seq=22 ttl=118 time=9.867 ms
64 bytes from 8.8.8.8: seq=34 ttl=118 time=15.462 ms
64 bytes from 8.8.8.8: seq=35 ttl=118 time=11.415 ms
64 bytes from 8.8.8.8: seq=40 ttl=118 time=13.381 ms
64 bytes from 8.8.8.8: seq=44 ttl=118 time=10.162 ms
64 bytes from 8.8.8.8: seq=49 ttl=118 time=15.912 ms
64 bytes from 8.8.8.8: seq=52 ttl=118 time=18.430 ms
64 bytes from 8.8.8.8: seq=53 ttl=118 time=10.238 ms
64 bytes from 8.8.8.8: seq=54 ttl=118 time=9.831 ms
64 bytes from 8.8.8.8: seq=59 ttl=118 time=10.645 ms
64 bytes from 8.8.8.8: seq=74 ttl=118 time=9.033 ms
64 bytes from 8.8.8.8: seq=82 ttl=118 time=7.420 ms
64 bytes from 8.8.8.8: seq=98 ttl=118 time=11.285 ms
64 bytes from 8.8.8.8: seq=110 ttl=118 time=13.781 ms
64 bytes from 8.8.8.8: seq=115 ttl=118 time=17.005 ms
64 bytes from 8.8.8.8: seq=123 ttl=118 time=11.826 ms
64 bytes from 8.8.8.8: seq=129 ttl=118 time=13.152 ms

```

tiagogaspar8 commented:
Hi @1715173329,
Can we get HW crypto enabled with this kernel upgrade by default?
Thanks.

1715173329 commented:
> Can we get HW crypto enabled with this kernel upgrade by default?

It's expected to be built as a module, maybe @antnyfls can look into this.
for reference: https://github.com/openwrt/openwrt/pull/15212#discussion_r1590546606

1715173329 commented:
> This PR runs on my Nano-Pi R6S, but performance is absolutely terrible. High packet loss which makes it essentially unusable. I had to roll back to previous 6.6 based image. See ping excerpt below. Nothing obvious in kernel log. Let me know if you need me to do any further testing.

I don't know how your cables are connected, maybe you can try to replace kmod-r8169 with kmod-r8125.

tiagogaspar8 commented:
> It's expected to be built as a module
Okok, I can also try to get a look at this, but when I investigated this, there was a patch done by nbd that brok the kmod compilation.

jmccrohan commented:
> > This PR runs on my Nano-Pi R6S, but performance is absolutely terrible. High packet loss which makes it essentially unusable. I had to roll back to previous 6.6 based image. See ping excerpt below. Nothing obvious in kernel log. Let me know if you need me to do any further testing.
> 
> I don't know how your cables are connected, maybe you can try to replace kmod-r8169 with kmod-r8125.

I've replaced r8169 with r8125 but issue persists. Existing 'good' 6.6 image was based on `r29345-f22d872c65`, so relatively recent. I will build another image based on main, excluding this PR, to confirm that it is the move to 6.12 that is causing this.

jmccrohan commented:
> I've replaced r8169 with r8125 but issue persists. Existing 'good' 6.6 image was based on `r29345-f22d872c65`, so relatively recent. I will build another image based on main, excluding this PR, to confirm that it is the move to 6.12 that is causing this.

@1715173329 Status update. Moved back to r8169, and created two images based off main, one with this PR applied and one without. The packetloss issue only appears in the 6.12 kernel image. The otherwise identical image running 6.6 does not display this issue. 

Let me know if there is anything in particular you require to assist with debugging. I can easily recreate and provide kernel logs etc.

1715173329 committed:
rockchip: Create kernel files for v6.12 (from v6.6)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>
Link: https://github.com/openwrt/openwrt/pull/18683
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

1715173329 committed:
rockchip: Restore kernel files for v6.6

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>
Link: https://github.com/openwrt/openwrt/pull/18683
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

PolynomialDivision commented:
Is this PR finished?

1715173329 commented:
> Is this PR finished?

Almost finished, except the issue reported by @tiagogaspar8. I don't have a RK3588 board to test now.
Cc @msylgj could you test this on NanoPC T6?

msylgj commented:
> > Is this PR finished?
> 
> Almost finished, except the issue reported by @tiagogaspar8. I don't have a RK3588 board to test now. Cc @msylgj could you test this on NanoPC T6?

simple test on NanoPC T6, kmod-8125, no packet loss. 
![image](https://github.com/user-attachments/assets/e74e77a3-cc9d-4945-8ef9-66604efdbcf4)

I'll use official repo test again in the weekend. @1715173329

tiagogaspar8 commented:
If I'm not mistaken, this was the patch that broke compilation of the driver:
https://github.com/openwrt/openwrt/blob/main/target/linux/generic/hack-6.12/260-crypto_test_dependencies.patch

1715173329 commented:
yes, see https://github.com/openwrt/openwrt/pull/15212#issuecomment-2095256568

linuxtextadventurer commented:
For whatever it's worth, I can confirm @jmccrohan findings on my NanoPi R6S as well. No matter which driver I'm using (r8129 vs r8125) there's a higher overall latency, latency spikes and even packet loss with 6.12. None of these issues exist in 6.6.

jmccrohan commented:
> For whatever it's worth, I can confirm @jmccrohan findings on my NanoPi R6S as well. No matter which driver I'm using (r8129 vs r8125) there's a higher overall latency, latency spikes and even packet loss with 6.12. None of these issues exist in 6.6.

Tagging Nano Pi R6S owners/previous contributors for visibility; @BWhitten @lespinasse @dmsza @PalebloodSky @Golui

Issue is as described in thread. Poor performance/packet loss when running 6.12 kernel.

PolynomialDivision commented:
> Hmmmm... Try to disable/enable flow offloading?

Wow, seems to fix it. :)

linuxtextadventurer commented:
> Hmmmm... Try to disable/enable flow offloading?

To me this seems a lot like a workaround to some underlying issue. Personally, I'd prefer to keep using SQM, which wouldn't work with flow offloading enabled, would it? Guess no upgrading to 6.12 for me then? 😞

jmccrohan commented:
> Hmmmm... Try to disable/enable flow offloading?

Issue persists irrespective of `tso`, `gso` and `gro` being on or off. Anything else you want me to check?

linuxtextadventurer commented:
> Issue persists irrespective of `tso`, `gso` and `gro` being on or off. Anything else you want me to check?

Oh, I thought @1715173329 meant this:
![Capture](https://github.com/user-attachments/assets/19f6f8ba-4466-4014-87c0-f1489ee8ed26)

(Personally, I haven't even tried that, because I still believe whatever issue is causing the packet loss and increase in latency absolutely needs to be fixed, even _if_ it could be worked around by fiddling with the flow offloading or any other settings.)

jmccrohan commented:
> > Issue persists irrespective of `tso`, `gso` and `gro` being on or off. Anything else you want me to check?
> 
> Oh, I thought @1715173329 meant this: 
> 
> (Personally, I haven't even tried that, because I still believe whatever issue is causing the packet loss and increase in latency absolutely needs to be fixed, even _if_ it could be worked around by fiddling with the flow offloading or any other settings.)

Good spot. I tested enabling hw flow offloading this evening and it made no difference (if anything it feels like it made things worse).

1715173329 commented:
Have you tried minimum build? Without any unnecessary configurations.
Default build works fine for us, I guess the issues are from other components and irrelevant to this PR itself.

Golui commented:
I'd like to rule out the obvious. Have you verified that this is not cased by an MTU mismatch?

hauke commented:
I would merge this PR even with the problems on RK3588. Kernel 66 is still the default and it is probably easier to fix this later.

1715173329 commented:
That symbol has been added into generic kconfig already.

jmccrohan commented:
> That symbol has been added into generic kconfig already.

Thanks. Missed that.

> Could you please verify that it is the communication between the client and the rockchip device that is the issue, as opposed to rockchip to the internet issue? Do not ping external services, do a direct ping/file copy to the OpenWRT device. Does the issue still arise?

The ping outputs show above are from the rockchip device itself to an external host. Pings from LAN devices behind the rockchip device to external internet hosts are similarly impacted. Pings been LAN devices and rockchip's LAN interface appear to be fine and responding normally.

This is running on a default image configuration with just the following files  `/etc/config/{dhcp,firewall,network,system}` copied from my existing good 6.6 image. I will work on paring these back even further, but made it is complicated by the fact that this is the household router.

jmccrohan commented:
I appreciate this PR has already been merged, but posting results of further testing for posterity. Let me know if you want me to split this into an issue instead.

Performance issues remain after I pared back `/etc/config`:

- Removed bridges and VLANs from network configuration to leave two basic 2.5G interfaces (WAN and LAN respectively).
- Disabled Firewall entirely
- Connected to LAN interface with statically IP'd laptop (bypassed trunked downstream switches/APs)

After multiple failed attempts (due to inability to resolve servers), I finally managed to complete a `speedtest-go` which indicates that this only impacts downloads. Uploads appear to unaffected.

```
root@OpenWrt:~# speedtest-go 

    speedtest-go v1.7.10 (git-dev) @showwin

✓ ISP: x.x.x.x (Virgin Media Ireland) [53.0000, -6.0000] 
✓ Found 20 Public Servers

✓ Test Server: [59135] 265.47km Manchester (United Kingdom) by Grain Connect
✓ Latency: 367.276146ms Jitter: 520.569377ms Min: 11.536ms Max: 1.800452209s
✓ Packet Loss Analyzer: Running in background (<= 30 Secs)
✓ Download: N/A (Used: 0.00MB) (Latency: 0ms Jitter: 0ms Min: 0ms Max: 0ms)
✓ Upload: 49.80 Mbps (Used: 64.20MB) (Latency: 906ms Jitter: 707ms Min: 15ms Max: 6955ms)
✓ Packet Loss: N/A
```

Issue #19130 seems to indicate a similar 6.12 issue on another arch.

jmccrohan commented:
It would appear this is a heisenbug. I used laptop and two USB r8152 ethernet dongles this evening to create a bridge and sniff the traffic between ISP modem and the rockchip router. Issue disappears while connected via the bridge, but immediately reappears when modem and router are connected directly.

As part of another test, I set up DHCP server connected to rockchip wan interface but issue similarly disappeared iperf3 been laptop and rockchip device is and is perfect line speed.

jmccrohan commented:
Sorry for the noise, but I've finally gotten to the bottom of this issue and am documenting the fix here.

Somewhere between 6.6 and 6.12, EEE (802.03az) support for 2.5Gb/s links was added to the kernel. This meant that 6.12 images had EEE enabled on my Rockchip WAN interface and started to trigger [this](https://www.ispreview.co.uk/index.php/2024/03/update-breaks-2-5gbps-port-on-virgin-media-hub-5-uk-broadband-router.html) [bug](https://community.virginmedia.com/discussions/Computers/hub-5-firmware-upgrade/5480079/replies/5483087) in my ISP provided Sagemcom F@st 3896 modem.

Performance issues disappeared once I either disabled EEE using `ethtool --set-eee eth1 eee off` or moved to one of the 1Gb/s links on the modem.


6.6 Kernel:
```
root@OpenWrt:~# ethtool --show-eee eth1
EEE settings for eth1:
	EEE status: enabled - inactive
	Tx LPI: 5 (us)
	Supported EEE link modes:  100baseT/Full
	                           1000baseT/Full
	Advertised EEE link modes:  100baseT/Full
	                            1000baseT/Full
	Link partner advertised EEE link modes:  100baseT/Full
	                                         1000baseT/Full
```

6.12 Kernel:
```
root@OpenWrt:~# ethtool --show-eee eth1
EEE settings for eth1:
	EEE status: enabled - active
	Tx LPI: 5 (us)
	Supported EEE link modes:  100baseT/Full
	                           1000baseT/Full
	                           2500baseT/Full
	Advertised EEE link modes:  100baseT/Full
	                            1000baseT/Full
	                            2500baseT/Full
	Link partner advertised EEE link modes:  100baseT/Full
	                                         1000baseT/Full
	                                         2500baseT/Full
```

1715173329 commented:
Good catch! Wonder if this also resolves @linuxtextadventurer's issue.
I plan to switch to 6.12 and drop 6.6 support.

jmccrohan commented:
> > disabled EEE using `ethtool --set-eee eth1 eee off`
> 
> Holy smokes! That's it! Thank you so, so much @jmccrohan for investigating and sharing your findings! Disabling EEE completely eliminated this issue for me, too. Oddly enough, I experienced this issue with a modem that doesn't even have a 2.5 Gbps port (using the DrayTekk Vigor167 I mentioned before). 🤔 In any case, I'm so incredibly glad this issue is fixed now!

Rather than messing around with `/etc/rc.local/`, the better fix is to add the `eee` option to the `eth1` device in `/etc/config/network` (option not exposed through LuCI).

```
config device
	option name 'eth1'
	option macaddr '01:02:03:04:05:06'
	option eee '0'
```
